### PR TITLE
Cucumber Expressions AST into Regex expansion

### PR DESCRIPTION
**Pull request by ilslv**
Moved from cucumber-rs/cucumber#154   
Part of cucumber-rs/cucumber#124

## Synopsis

This `PR` implements `Cucumber expressions` `AST` into `Regex` expansion. 

## Solution

This is done with `IntoRegexCharIter` trait to avoid unnecessary overhead. 

## Checklist

- Created PR:
    - [x] In [draft mode][l:1]
    - [x] Name contains `Draft: ` prefix
    - [x] Name contains issue reference
    - [x] Has assignee
- [x] Documentation is updated (if required)
- [x] Tests are updated (if required)
- [x] Changes conform code style
- [x] CHANGELOG entry is added (if required)
- [x] FCM (final commit message) is posted
    - [x] and approved
- [x] [Review][l:2] is completed and changes are approved
- Before merge:
    - [x] Milestone is set
    - [x] PR's name and description are correct and up-to-date
    - [x] `Draft: ` prefix is removed
    - [x] All temporary labels are removed

[l:1]: https://help.github.com/en/articles/about-pull-requests#draft-pull-requests
[l:2]: https://help.github.com/en/articles/reviewing-changes-in-pull-requests

### CODE REVIEW

**Comment by ilslv** [comment]
`derive_more` crate is dependent on `syn = 1`. That why after running `cargo +nightly update -Z minimal-versions` `syn` gets downgraded to `1.0`, where there is no `Path::get_ident()` method. 

UPD: Waiting for https://github.com/JelteF/derive_more/pull/175

**Comment by ilslv** [comment]
## FCM

```
Cucumber Expressions AST into `Regex` expansion (#2)
```

File: src/expand/mod.rs [hunk 1]
```diff
@@ -0,0 +1,622 @@
...
+    /// Transforms `Input` into [`Regex`] with `Parameters` in addition to
+    /// the [default ones][0]
+    ///
+    /// # Errors
+    ///
+    /// See [`Error`] for more details.
+    ///
+    /// [0]: Expression#parameter-types
+    /// [`Error`]: ExpansionError
+    pub fn regex_with_parameters<Input, Parameters>(
```
Comment (tyranron) [inline src/expand/mod.rs:106]: It's hard to grasp the whole idea here. Can we provide a meaningful example?

File: src/expand/mod.rs [hunk 2]
```diff
@@ -0,0 +1,622 @@
...
+        iter::FlatMap<
+            vec::IntoIter<SingleExpression<Input>>,
+            <SingleExpression<Input> as IntoRegexCharIter<Input>>::Iter,
+            fn(
+                SingleExpression<Input>,
+            )
+                -> <SingleExpression<Input> as IntoRegexCharIter<Input>>::Iter,
+        >,
+    >,
+    iter::Once<Result<char, UnknownParameterError<Input>>>,
```
Comment (tyranron) [inline src/expand/mod.rs:221]: Once we have TAIT we won't need this aliases anymore. It's worth mentioning with TODO comment for future.

File: src/expand/mod.rs [hunk 3]
```diff
@@ -0,0 +1,622 @@
...
+//                     Kai Ren <tyranron@gmail.com>
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! [Cucumber Expressions][0] [AST] into [`Regex`] transformation
+//! definitions.
```
Comment (tyranron) [inline src/expand/mod.rs]: We also need to mention somewhere [upstream production rules](https://github.com/cucumber/cucumber-expressions/blob/main/ARCHITECTURE.md#production-rules).

If we align with them 1-to-1, then they should be referred in transformation functions docs directly.

File: src/lib.rs [hunk 4]
```diff
@@ -85,3 +95,24 @@
...
+
+#[doc(inline)]
+pub use self::ast::{
+    Alternation, Alternative, Expression, Optional, Parameter,
+    SingleAlternation, SingleExpression, Spanned,
+};
+#[cfg(feature = "expand-into-regex")]
+#[doc(inline)]
+pub use self::expand::{
+    ExpansionError as Error, IntoRegexCharIter, ParametersProvider,
```
Comment (tyranron) [inline src/lib.rs]: I think we can omit re-exporting feature in root. They're quite specific to live in their on module.

File: src/expand/mod.rs [hunk 5]
```diff
@@ -0,0 +1,622 @@
...
+use either::Either;
+use nom::{AsChar, InputIter};
+use regex::Regex;
+
+use crate::{
+    parse, Alternation, Alternative, Expression, Optional, Parameter,
+    SingleAlternation, SingleExpression, Spanned,
+};
+
+pub use with_parameters::{ParametersProvider, WithParameters};
```
Comment (tyranron) [inline src/expand/mod.rs]: Could it be?
```rust
pub mod parameters;

pub use parameters::{Provider as ParametersProvider, WithCustom as WithCustomParameters};
```

**Review by tyranron — CHANGES_REQUESTED** [review]
@ilslv mostly OK, just bikeshedding is required.

**Review by tyranron — APPROVED** [review]

**Pull request merged**